### PR TITLE
[Bugfix][Attention] Normalize FlashInfer prefill LSE before merging

### PR DESCRIPTION
**Pull request by yimdev**
## Purpose

Prefix-cache hits could produce different logits from a cold prefill and
eventually lead to repetitive output. This was initially observed with GLM-5.2
in a disaggregated prefill/decode deployment.

KV cache digests matched before and after transfer, and the issue also
reproduced with BF16 KV cache. A local prefix-cache cold/warm comparison then
reproduced the logits divergence without KV transfer, narrowing the issue to
the cached-context attention path.

A cold prefill computes attention in one operation. On a prefix-cache hit,
vLLM computes attention over the cached context and new suffix separately, then
combines the two partial states with `merge_attn_states`.

FlashInfer prefill kernels, including the TRT-LLM ragged kernel, return base-2
log-sum-exp values. vLLM's CUDA and Triton `merge_attn_states` implementations
use natural `exp` and `log`. Passing base-2 LSE values directly therefore
produces incorrect context/suffix weights.

This PR:

- adds a shared `log2_lse_to_ln` conversion helper;
- converts LSE returned by the FlashInfer MLA prefill backend;
- converts LSE returned by the TRT-LLM ragged MLA prefill backend;
- converts generic FlashInfer DCP prefill LSE after the base-2 cross-rank
  reduction and before the natural-log attention-state merge.

The conversion is performed at the external kernel boundary so the rest of
vLLM continues to use the natural-log LSE contract.

### Duplicate-work check

No open PR currently fixes the MLA prefill paths changed here.

PR #36503 previously experimented with scaling generic FlashInfer prefill LSE
in an older commit (`f10579e`), but that change is not present in its current
diff. Its current scope is persistent buffer management for FlashInfer DCP
decode. This PR fixes the missing conversions on current main and additionally
covers both FlashInfer MLA prefill and TRT-LLM ragged MLA prefill.

PR #47779 concerns SM120 sparse MLA decode DCP and does not modify these
prefill merge paths.

## Test Plan

```bash
.venv/bin/python -m pytest -q -rs \
  tests/v1/attention/test_mla_backends.py \
  -k 'test_chunked_context_backend_correctness and \
      (FLASHINFER or TRTLLM_RAGGED)'

.venv/bin/pre-commit run --files \
  vllm/v1/attention/backends/utils.py \
  vllm/v1/attention/backends/flashinfer.py \
  vllm/v1/attention/backends/mla/prefill/flashinfer.py \
  vllm/v1/attention/backends/mla/prefill/trtllm_ragged.py

git diff --check
```

Before/after error measurements used the same B300 GPU, test selection, and
random sequence. The old behavior was simulated by replacing the conversion
helper with an identity function at runtime.

## Test Result

Affected auto/FP8 matrix:

- 6 passed
- 2 skipped because FlashInfer does not support the tested GLM dimensions

Maximum absolute output error against the SDPA reference:

| Prefill path | Before | After |
| --- | ---: | ---: |
| FlashInfer / DeepSeek | 0.236328 | 0.0078125 |
| TRT-LLM ragged / DeepSeek | 0.228516 | 0.0078125 |
| TRT-LLM ragged / GLM | 0.265625 | 0.0078125 |

The results were identical across CUTLASS_MLA, FLASHINFER_MLA, and TRITON_MLA,
which is consistent with the error originating in the shared context/suffix
merge.

Pre-commit and `git diff --check` passed.

The complete `test_chunked_context_backend_correctness` matrix produced 10
passed, 8 skipped, and 2 failures. Both failures are in the unchanged
FlashAttention FA4 path because the local extension does not accept the
`fp8_kv_dequant` keyword. They occur before LSE merging.

### AI assistance

AI assistance was used.

---
<details>
<summary> Essential Elements of an Effective PR Description Checklist </summary>

- [x] The purpose of the PR is described.
- [x] The test plan and commands are included.
- [x] Local test results and the environment-specific failures are included.
- [ ] Model evaluation results must be added before submission.

</details>

### CODE REVIEW

**Comment by MatthewBonanni** [comment]
Thanks for catching this! Can we tighten the tolerance of `_run_backend_correctness` now?

**Comment by yimdev** [comment]
> Thanks for catching this! Can we tighten the tolerance of `_run_backend_correctness` now?

Thanks for the review. I tested stricter tolerances on an NVIDIA B300 before choosing atol=0.15 with the existing rtol=0.01.
Test results:
- test_chunked_context_backend_correctness with the final atol=0.15: 12 passed, 8 skipped.
- The fixed implementation had a maximum absolute error of 0.0078125 in the affected chunked-context merge cases.
- I also ran the complete locally supported test_backend_correctness matrix with atol=0.1: 1718 passed, 1152 skipped, 10 failed.
- All 10 failures were FP8/FP8-E4M3 small_decode or mixed_small cases:
  - FlashAttention prefill: 1
  - FlashInfer prefill: 4
  - TRT-LLM ragged prefill: 3
  - TokenSpeed MLA prefill: 2
- The maximum valid absolute error observed in that run was 0.111328.
- Re-running the 10 failed cases resulted in 9 passing and 1 failing, with the remaining failure at 0.105469. This variation occurs because the test seeds PyTorch once at module import rather than once per parametrized case, so test ordering and CI sharding can produce different inputs.
To ensure the relaxed tolerance still detects this regression, I simulated the old behavior by bypassing the log2-to-natural-log LSE conversion. With atol=0.15, all three locally runnable affected configurations still failed, with maximum absolute errors of:
- FlashInfer/DeepSeek: 0.236328
- TRT-LLM ragged/DeepSeek: 0.228516
- TRT-LLM ragged/GLM: 0.265625

Finally, I selected 0.15. Let's try a full CI run.

**Comment by MatthewBonanni** [comment]
Thanks! Can you make the threshold based on dtype? It would be nice to tighten it more for bf16; I suspect 0.15 still gives a lot of opportunities for correctness issues to slip through

**Comment by yimdev** [comment]
> Thanks! Can you make the threshold based on dtype? It would be nice to tighten it more for bf16; I suspect 0.15 still gives a lot of opportunities for correctness issues to slip through

Sure. threshold is based on dtype now.

**Review by MatthewBonanni — APPROVED** [review]
LGTM, thanks!
